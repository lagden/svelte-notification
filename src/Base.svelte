<script>
	import {fade, slide} from 'svelte/transition'

	export let mode
</script>

<div
	role="status"
	aria-live="polite"
	class="_tadashi_svelte_notification"
	class:_tadashi_svelte_notification__success={mode === 'success' || mode === 'ok'}
	class:_tadashi_svelte_notification__warning={mode === 'warning' || mode === 'warn'}
	class:_tadashi_svelte_notification__danger={mode === 'danger' || mode === 'error'}
	class:_tadashi_svelte_notification__info={mode === 'info'}
	in:fade|local
	out:slide|local
>
	<div class="_tadashi_svelte_notification__content">
		<slot />
	</div>
	<button
		aria-label="delete notification"
		class="_tadashi_svelte_notification__btn"
		on:click
	>Ã—</button>
</div>

<style>
	:root {
		--tadashi_svelte_notification_margin: 0 0 0.9em;
		--tadashi_svelte_notification_background: hsl(0deg 100% 100% / 90%);
		--tadashi_svelte_notification_color: hsl(0deg 0% 0%);
		--tadashi_svelte_notification_border_radius: 3px;
		--tadashi_svelte_notification_box_shadow: 0 4px 10px hsl(0deg 0% 0% / 10%);

		--tadashi_svelte_notification__content_padding: 0.9em;

		--tadashi_svelte_notification__btn_font_size: 1.5em;
		--tadashi_svelte_notification__btn_font_family: monospace;

		--tadashi_svelte_notification__success_background: hsl(149deg 96% 42% / 90%);
		--tadashi_svelte_notification__success_color: hsl(0deg 100% 100%);

		--tadashi_svelte_notification__warning_background: hsl(44deg 100% 50%  / 90%);
		--tadashi_svelte_notification__warning_color: hsl(0deg 0% 0%);

		--tadashi_svelte_notification__danger_background: hsl(359deg 88% 64% / 90%);
		--tadashi_svelte_notification__danger_color: hsl(0deg 100% 100%);

		--tadashi_svelte_notification__info_background: hsl(213deg 56% 47% / 90%);
		--tadashi_svelte_notification__info_color: hsl(0deg 100% 100%);
	}

	._tadashi_svelte_notification {
		position: relative;
		display: grid;
		grid-template-columns: 1fr 40px;
		grid-gap: 1px;
		margin: var(--tadashi_svelte_notification_margin);
		border-radius: var(--tadashi_svelte_notification_border_radius);
		box-shadow: var(--tadashi_svelte_notification_box_shadow);
		overflow: hidden;
	}

	._tadashi_svelte_notification__content {
		padding: var(--tadashi_svelte_notification__content_padding);
		box-sizing: border-box;
		word-wrap: break-word;
	}

	._tadashi_svelte_notification__btn {
		padding: 0;
		margin: 0;
		border: none;
		border-radius: 0;
		outline: none;
		background: none;
		cursor: pointer;
		box-sizing: border-box;
		font-size: var(--tadashi_svelte_notification__btn_font_size);
		font-family: var(--tadashi_svelte_notification__btn_font_family);
	}

	._tadashi_svelte_notification > ._tadashi_svelte_notification__content,
	._tadashi_svelte_notification > ._tadashi_svelte_notification__btn {
		background: var(--tadashi_svelte_notification_background);
		color: var(--tadashi_svelte_notification_color);
	}

	._tadashi_svelte_notification__success > ._tadashi_svelte_notification__content,
	._tadashi_svelte_notification__success > ._tadashi_svelte_notification__btn {
		background: var(--tadashi_svelte_notification__success_background);
		color: var(--tadashi_svelte_notification__success_color);
	}

	._tadashi_svelte_notification__warning > ._tadashi_svelte_notification__content,
	._tadashi_svelte_notification__warning > ._tadashi_svelte_notification__btn {
		background: var(--tadashi_svelte_notification__warning_background);
		color: var(--tadashi_svelte_notification__warning_color);
	}

	._tadashi_svelte_notification__danger > ._tadashi_svelte_notification__content,
	._tadashi_svelte_notification__danger > ._tadashi_svelte_notification__btn {
		background: var(--tadashi_svelte_notification__danger_background);
		color: var(--tadashi_svelte_notification__danger_color);
	}

	._tadashi_svelte_notification__info > ._tadashi_svelte_notification__content,
	._tadashi_svelte_notification__info > ._tadashi_svelte_notification__btn {
		background: var(--tadashi_svelte_notification__info_background);
		color: var(--tadashi_svelte_notification__info_color);
	}
</style>
