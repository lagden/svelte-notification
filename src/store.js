import {writable, get} from 'svelte/store'

const store = writable(new Set([]))
const getData = store => get(store)

export {
	store,
	getData
}
