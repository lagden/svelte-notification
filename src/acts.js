import {store} from './store'

let cc = 0
function hexId() {
	return Number(cc++).toString(26)
}

const base = {
	mode: 'normal',
	message: '',
	lifetime: false
}

export function add(notification) {
	notification = {
		...base,
		...notification,
		id: `tadashi_svelte_notification_${hexId()}`
	}
	store.update(n => {
		n.add(notification)
		return n
	})
}

export function remove(notification) {
	store.update(n => {
		n.delete(notification)
		return n
	})
}
