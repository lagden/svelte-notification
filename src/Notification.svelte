<script>
	import {onDestroy} from 'svelte'
	import {remove} from './acts'

	import Base from './Base.svelte'

	export let notification

	const {
		mode = 'normal',
		message = '',
		lifetime = false
	} = notification

	let _timeout

	if (lifetime) {
		let _time = Number.isNaN(lifetime) ? 1000 : Number(lifetime) * 1000
		_timeout = setTimeout(() => {
			remove(notification)
		}, _time)
	}

	function onRemoveNotification(notification) {
		remove(notification)
	}

	onDestroy(() => {
		if (_timeout) {
			clearTimeout(_timeout)
		}
	})
</script>

<Base {mode} on:click={onRemoveNotification(notification)}>{@html message}</Base>
