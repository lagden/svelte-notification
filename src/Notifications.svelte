<script>
	import {fade} from 'svelte/transition'
	import {store} from './store'

	import Notification from './Notification.svelte'

	function init(node) {
		document.body.insertAdjacentElement('beforeend', node)
	}
</script>

{#if $store.size > 0}
	<div
		class="_tadashi_svelte_notifications"
		use:init
		in:fade
		out:fade
	>
		{#each [...$store] as notification (notification.id)}
			<Notification {notification} />
		{/each}
	</div>
{/if}

<style>
:root {
	--tadashi_svelte_notifications_width: 270px;
	--tadashi_svelte_notifications_margin: 0.9em;
	--tadashi_svelte_notifications_zindex: 1001;
}

._tadashi_svelte_notifications {
	position: fixed;
	width: var(--tadashi_svelte_notifications_width);
	margin: var(--tadashi_svelte_notifications_margin);
	top: 0;
	right: 0;
	z-index: var(--tadashi_svelte_notifications_zindex);
}
</style>
