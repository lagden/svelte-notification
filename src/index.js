/* istanbul ignore file */

import * as acts from './acts'
import Notifications from './Notifications.svelte'

export {
	Notifications,
	acts
}
