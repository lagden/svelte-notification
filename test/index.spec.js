/* globals describe, afterEach, test, expect */

import {cleanup, render} from '@testing-library/svelte'
import {Notifications, acts} from '../src'

describe('Notifications', () => {
	afterEach(cleanup)

	test('should match snapshot', done => {
		const {container} = render(Notifications)
		acts.add({message: 'test'})

		setTimeout(() => {
			expect(container).toMatchSnapshot()
			done()
		}, 1000)
	})
})
