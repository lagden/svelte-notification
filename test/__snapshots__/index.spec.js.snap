// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications should match snapshot 1`] = `
<body>
  <div>
    
  </div>
  <div
    class="_tadashi_svelte_notifications svelte-1cix5mn"
    style=""
  >
    
    <div
      aria-live="polite"
      class="_tadashi_svelte_notification svelte-1aia55s"
      role="status"
    >
      <div
        class="_tadashi_svelte_notification__content svelte-1aia55s"
      >
        test
        
      </div>
       
      <button
        aria-label="delete notification"
        class="_tadashi_svelte_notification__btn svelte-1aia55s"
      >
        Ã—
      </button>
    </div>
  </div>
</body>
`;
